### PR TITLE
CRM-17606 - Support case tokens when creating PDF

### PR DESCRIPTION
**Pull request by colemanw**
- [CRM-17606: Create Document \/ Letter from Case(s) - with case tokens](https://issues.civicrm.org/jira/browse/CRM-17606)

### CODE REVIEW

File: api/v3/Case.php [hunk 1]
```diff
@@ -416,6 +416,9 @@ function _civicrm_api3_case_read($caseId, $options) {
   if ($dao->find(TRUE)) {
     $case = array();
     _civicrm_api3_object_to_array($dao, $case);
+
+    _civicrm_api3_custom_data_get($case, 'Case', $caseId);
```
Comment (eileenmcnaughton) [inline api/v3/Case.php]: This is to save us from using the get api :-)

File: api/v3/utils.php [hunk 2]
```diff
@@ -2602,7 +2602,7 @@ function _civicrm_api3_api_resolve_alias($entity, $fieldName, $action = 'create'
     return $meta[$fieldName]['name'];
   }
   foreach ($meta as $info) {
-    if ($fieldName == CRM_Utils_Array::value('uniqueName', $info)) {
+    if ($fieldName == $info['name'] || $fieldName == CRM_Utils_Array::value('uniqueName', $info)) {
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: The CRM_Utils_Array::value emerges unscathed....

**Comment by eileenmcnaughton** [comment]
Looks good ...

File: api/v3/Case.php (hunk 1 above)
Comment (colemanw) [inline api/v3/Case.php]: ??? This *is* the get api!

File: api/v3/utils.php (hunk 2 above)
Comment (colemanw) [inline api/v3/utils.php]: `$info['uniqueName']` is not always set.
`CRM_Utils_Array::value` *is* useful for some things, I'll admint.

**Comment by colemanw** [comment]
Thanks. When I first dug into this I was dismayed that `CRM_Utils_Token` is such a mess.
The new caseTokens I've added don't completely break the mold but they are at least 100% reliant on the api. With the addition of some tests, more entities could be converted to use the generic `getApiTokenReplacement` function to start detangling things.

File: api/v3/Case.php (hunk 1 above)
Comment (eileenmcnaughton) [inline api/v3/Case.php]: Then I don't know why that was required.... The syntax conformance test already tests get & create for all entities defined in the static map (I thought).

**Comment by eileenmcnaughton** [comment]
I think you mispelt mould....

**Pull request merged**